### PR TITLE
Modify the fastboot and MM plugins used to upgrade fibocom fm101 and the fwupdmgr reinstall logic

### PR DESCRIPTION
**Pull request by cihun**
Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
  - [ ] Fill out `README.md` with update protocol
  - [ ] Fill out `README.md` with any custom quirks and flags
  - [x] Fill out `README.md` with the vendor ID security value
  - [ ] Implement `FuFirmware->write()` and include at least one fuzzer testcase in `src/fuzzing/firmware` for any custom `FuFirmware` subclass
  - [ ] CI run of the plugin for at least one target
  - [ ] Document targets CI isn't currently run and the reasons (i.e. minimum versions needed or distribution limitations etc).
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/fastboot/fastboot.quirk [hunk 1]
```diff
@@ -6,3 +6,10 @@ Plugin = fastboot
 [USB\VID_18D1&PID_D00D]
 FastbootBlockSize = 16384
 FastbootOperationDelay = 250
+
+# Fibocom device fastboot
+[USB\VID_2CB7&PID_D00D]
+Plugin = fastboot
+FastbootBlockSize = 16384
+FastbootOperationDelay = 0
+CounterpartGuid = USB\VID_2CB7&PID_01A2
```
Comment (hughsie) [inline plugins/fastboot/fastboot.quirk:15]: i think you need a newline at the end of this file

File: plugins/modem-manager/modem-manager.quirk [hunk 2]
```diff
@@ -73,3 +73,14 @@ ModemManagerFirehoseProgFile = prog_firehose_sdx24.mbn
...
+[USB\VID_2CB7&PID_01A2]
+Summary = Fibocom Fibocom FM101-GL Module
+CounterpartGuid = USB\VID_2CB7&PID_D00D
+ModemManagerFirehoseProgFile = firehose-prog.mbn
+Flags = detach-at-fastboot-has-no-response
+
+# Fibocom device fastboot
+[USB\VID_2CB7&PID_D00D]
+Summary = Fibocom Fibocom FM101 Module
+CounterpartGuid = USB\VID_2CB7&PID_01A2
```
Comment (hughsie) [inline plugins/modem-manager/modem-manager.quirk:86]: also needs a newline

File: plugins/fastboot/fu-fastboot-device.c [hunk 3]
```diff
@@ -24,6 +24,8 @@ struct _FuFastbootDevice {
 	gboolean secure;
 	guint blocksz;
 	guint operation_delay;
+
+	int fibocom_flash_end; //fibocom modem needs to be marked When is flash finished
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:28]: comments all need to be `/* foo bar */` with a space between the stars

File: plugins/fastboot/fu-fastboot-device.c [hunk 4]
```diff
@@ -340,6 +342,9 @@ fu_fastboot_device_setup(FuDevice *device, GError **error)
 	if (secure != NULL && secure[0] != '\0')
 		self->secure = TRUE;
 
+	if (fu_device_has_vendor_id(self, "USB:0x2CB7")) {
+		self->fibocom_flash_end = 0;
+	}
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:347]: ALL fibocom devices?!

File: plugins/fastboot/fu-fastboot-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/fastboot/fu-fastboot-device.c:28]: also guint please

File: plugins/fastboot/fu-fastboot-device.c [hunk 5]
```diff
@@ -623,6 +628,21 @@ fu_fastboot_device_write_firmware(FuDevice *device,
 				  GError **error)
 {
 	g_autoptr(FuFirmware) manifest = NULL;
+	g_autoptr(FuFirmware) flash_end_file = NULL;
+	FuFastbootDevice *self = FU_FASTBOOT_DEVICE(device);
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:632]: self is normally first, i.e. before the g_autoptr() lines

File: plugins/fastboot/fu-fastboot-device.c [hunk 6]
```diff
@@ -623,6 +628,21 @@ fu_fastboot_device_write_firmware(FuDevice *device,
...
+
+	/* fibocom modem Set the fibocom_flash_end value */
+	if (fu_device_has_vendor_id(self, "USB:0x2CB7")) {
+		flash_end_file = fu_firmware_get_image_by_id(firmware, "flash_end", NULL);
+		manifest = fu_firmware_get_image_by_id(firmware, "partition_nand.xml", NULL);
+		if (flash_end_file != NULL && manifest == NULL) {
+			self->fibocom_flash_end-=1;
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:639]: formatting is funky -- have you run ./contrib/setup and set up precommit? clang-format should have fixed this..

File: plugins/modem-manager/fu-mm-device.c [hunk 7]
```diff
@@ -170,6 +170,340 @@ validate_firmware_update_method(MMModemFirmwareUpdateMethod methods, GError **er
...
+static gboolean
+fu_mm_device_at_cmd_get_version_fibocom(FuDevice *device, const gchar *cmd, const gchar *regex_char, gchar **version)
+{
+	// The fibocom modem obtains the device version through AT
+	g_autoptr(GString) reply = g_string_new("");
+	g_autoptr(GRegex) regex = NULL;
+	g_autoptr(GError) error_regex = NULL;
+	gchar *word = NULL;
+	GMatchInfo *match_info;
+	for (int i = 0;i < 3;i++){
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:262]: I think you want `fu_device_retry` for this

File: plugins/modem-manager/fu-mm-device.c [hunk 8]
```diff
@@ -170,6 +170,340 @@ validate_firmware_update_method(MMModemFirmwareUpdateMethod methods, GError **er
...
+fu_mm_device_io_close(FuMmDevice *self, GError **error);
+
+static gboolean
+fu_mm_device_at_cmd_get_version_fibocom(FuDevice *device, const gchar *cmd, const gchar *regex_char, gchar **version)
+{
+	// The fibocom modem obtains the device version through AT
+	g_autoptr(GString) reply = g_string_new("");
+	g_autoptr(GRegex) regex = NULL;
+	g_autoptr(GError) error_regex = NULL;
+	gchar *word = NULL;
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:260]: you can use g_autofree here

File: plugins/modem-manager/fu-mm-device.c [hunk 9]
```diff
@@ -170,6 +170,340 @@ validate_firmware_update_method(MMModemFirmwareUpdateMethod methods, GError **er
...
+		*version = g_strdup("NONE");
+		g_debug("Not version");
+		return FALSE;
+	}
+
+	regex = g_regex_new(regex_char, 0, 0, &error_regex);
+	if (regex == NULL) {
+		g_debug("%s", error_regex->message);
+	} else {
+		g_regex_match(regex, reply->str, 0, &match_info);
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:289]: what's the string format of this?

File: plugins/modem-manager/fu-mm-device.c [hunk 10]
```diff
@@ -170,6 +170,340 @@ validate_firmware_update_method(MMModemFirmwareUpdateMethod methods, GError **er
...
+	}
+
+	for (guint i = 0; i < n_ports; i++) {
+		if (ports[i].type == MM_MODEM_PORT_TYPE_AT) {
+			self->port_at = g_strdup_printf("/dev/%s", ports[i].name);
+			break;
+		}
+	}
+	if (self->port_at == NULL){
+		self->port_at = g_strdup_printf("/dev/ttyUSB1");
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:401]: why?

File: plugins/modem-manager/fu-mm-device.c [hunk 11]
```diff
@@ -189,6 +523,10 @@ fu_mm_device_probe_default(FuDevice *device, GError **error)
 	 * be the device sysfs path or a different user-provided id */
 	self->inhibition_uid = mm_modem_dup_device(modem);
 
+	if (g_strcmp0("Fibocom Wireless Inc.", mm_modem_get_manufacturer(modem)) == 0) {
+		return fu_mm_device_probe_fibocom(device, error);
+	}
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:528]: is there no VID to use here?

File: src/fu-util.c [hunk 12]
```diff
@@ -2870,10 +2911,14 @@ fu_util_reinstall(FuUtilPrivate *priv, gchar **values, GError **error)
...
-	if (rel == NULL)
-		return FALSE;
+	rel = 
+		fu_util_get_backup_release_for_device_version(priv, dev, fwupd_device_get_version(dev), error);
+	if (rel == NULL) {
+		rel =
+			fu_util_get_release_for_device_version(priv, dev, fwupd_device_get_version(dev), error);
+		if (rel == NULL)
+			return FALSE;
+	}
```
Comment (hughsie) [inline src/fu-util.c:2921]: why? Shouldn't the backup be found exactly like the other releases?

File: src/fu-util.c [hunk 13]
```diff
@@ -2870,10 +2911,18 @@ fu_util_reinstall(FuUtilPrivate *priv, gchar **values, GError **error)
...
 	/* try to lookup/match release from client */
-	rel =
-	    fu_util_get_release_for_device_version(priv, dev, fwupd_device_get_version(dev), error);
-	if (rel == NULL)
-		return FALSE;
+	rel = fu_util_get_backup_release_for_device_version(priv,
+							    dev,
+							    fwupd_device_get_version(dev),
+							    error);
+	if (rel == NULL) {
```
Comment (hughsie) [inline src/fu-util.c:2918]: i'm still confused what bug you're working around; is it that you want the backup remote to have a higher priority?

File: plugins/modem-manager/fu-mm-device.c [hunk 14]
```diff
@@ -649,6 +1048,21 @@ fu_mm_device_at_cmd_cb(FuDevice *device, gpointer user_data, GError **error)
 	}
 	fu_dump_bytes(G_LOG_DOMAIN, "read", at_res);
 	buf = g_bytes_get_data(at_res, &bufsz);
+
+	// The first time the modem returns may be the command itself with one \n missing.
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:1052]: /* comments please */ and lower case single line ones at that

File: plugins/modem-manager/fu-mm-device.c [hunk 15]
```diff
@@ -170,6 +170,401 @@ validate_firmware_update_method(MMModemFirmwareUpdateMethod methods, GError **er
...
+	g_autoptr(GError) error_sim = NULL;
+	const gchar *regex_list[] = {
+	    "\\d{5}.\\d{4}.\\d{2}.\\d{2}.\\d{2}.\\d{2}_\\d{4}.\\d{4}.\\d{3}.\\d{3}.\\d{3}_\\w\\d{"
+	    "2}",
+	    "\\d{5}.\\d{4}.\\d{2}.\\d{2}.\\d{2}.\\d{2}",
+	    "\\w\\d{2}",
+	    "\\d{4}.\\d{4}.\\d{3}",
+	    "\\d{3}.\\d{3}",
+	    "\\d{4}.\\d{4}.\\d{4}.\\d{4}_\\d{2}.\\d{2}",
+	};
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-device.c:380]: why so many formats?

File: plugins/modem-manager/fu-mm-device.c (hunk 9 above)
Comment (hughsie) [reply plugins/modem-manager/fu-mm-device.c:307]: this isn't really resolved

File: plugins/fastboot/fu-fastboot-device.c [hunk 16]
```diff
@@ -622,7 +627,21 @@ fu_fastboot_device_write_firmware(FuDevice *device,
...
+
+	/* fibocom modem Set the fibocom_flash_end value */
+	if (fu_device_has_vendor_id(self, "USB:0x2CB7")) {
+		flash_end_file = fu_firmware_get_image_by_id(firmware, "flash_end", NULL);
+		manifest = fu_firmware_get_image_by_id(firmware, "partition_nand.xml", NULL);
+		if (flash_end_file != NULL && manifest == NULL) {
+			self->fibocom_flash_end -= 1;
+			return TRUE;
+		} else if (flash_end_file == NULL && manifest != NULL) {
+			self->fibocom_flash_end += 1;
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:642]: i don't understand what all these fibocom_flash_end values are.

**Comment by cihun** [comment]
Hi,

What other questions do I need to answer?

Thanks.

**Comment by hughsie** [comment]
> What other questions do I need to answer?

The review comments on https://github.com/fwupd/fwupd/pull/5729 please.

**Comment by cihun** [comment]
Hi @hughsie ,

I responded to all the comments.

Please check, if there is no problem I will Resolve conversation.

Thanks.

**Comment by hughsie** [comment]
There are 3 open questions on this PR.

File: src/fu-util.c (hunk 13 above)
Comment (hughsie) [reply src/fu-util.c:2918]: Does this fix the problem properly: https://github.com/fwupd/fwupd/pull/5769 -- this should mean the workaround is no longer needed

File: plugins/fastboot/fu-fastboot-device.c [hunk 17]
```diff
@@ -24,6 +24,9 @@ struct _FuFastbootDevice {
 	gboolean secure;
 	guint blocksz;
 	guint operation_delay;
+
+	/* fibocom modem needs to be marked When is flash finished */
+	int fibocom_flash_end;
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:29]: could this be a boolean?

File: plugins/fastboot/fu-fastboot-device.c [hunk 18]
```diff
@@ -668,6 +688,14 @@ fu_fastboot_device_set_quirk_kv(FuDevice *device,
 static gboolean
 fu_fastboot_device_attach(FuDevice *device, FuProgress *progress, GError **error)
 {
+	/* if flash_end is greater than zero, flash is not finished and attach cannot be executed */
+	FuFastbootDevice *self = FU_FASTBOOT_DEVICE(device);
+	if (fu_device_has_vendor_id(self, "USB:0x2CB7")) {
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-device.c:693]: i think this should be a private flag and set in probe() or setup()

File: plugins/fastboot/fu-fastboot-plugin.c [hunk 19]
```diff
@@ -30,9 +30,18 @@ fu_fastboot_plugin_constructed(GObject *obj)
 	fu_plugin_add_device_gtype(plugin, FU_TYPE_FASTBOOT_DEVICE);
 }
 
+static void
+fu_fastboot_plugin_device_added(FuPlugin *self, FuDevice *device)
+{
+	if (fu_device_has_vendor_id(device, "USB:0x2CB7")) {
+		fu_device_set_plugin(device, "fibocom");
+	}
```
Comment (hughsie) [inline plugins/fastboot/fu-fastboot-plugin.c:38]: this needs to be done from the quirk file

File: plugins/fibocom/fu-fibocom-plugin.c [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+	flash_end_file = fu_firmware_get_image_by_id(firmware, "flash_end", NULL);
+	manifest = fu_firmware_get_image_by_id(firmware, "partition_nand.xml", NULL);
+
+	if (flash_end_file != NULL && manifest == NULL) {
+		self->fibocom_flash_end -= 1;
+		g_info("fibocom flash end number: %d", self->fibocom_flash_end);
+		return TRUE;
+	} else if (flash_end_file == NULL && manifest != NULL) {
+		self->fibocom_flash_end += 1;
+	}
```
Comment (hughsie) [inline plugins/fibocom/fu-fibocom-plugin.c:50]: why an integer? can you explain what it's doing please?

File: plugins/modem-manager/fu-mm-plugin.c [hunk 21]
```diff
@@ -218,6 +224,11 @@ fu_mm_plugin_inhibit_device(FuPlugin *plugin, FuDevice *device, GError **error)
 	FuModemManagerPlugin *self = FU_MODEM_MANAGER_PLUGIN(plugin);
 	g_autoptr(FuMmDevice) shadow_device = NULL;
 
+	self->uninhibit_flag = FALSE;
+	if (fu_device_has_vendor_id(FU_MM_DEVICE(device), "USB:0x2CB7")) {
+		self->uninhibit_flag = TRUE;
+	}
```
Comment (hughsie) [inline plugins/modem-manager/fu-mm-plugin.c:230]: why? I'm very confused what this is trying to do

**Pull request closed without merging**